### PR TITLE
Implement Node and Graph modules in OCaml

### PR DESCRIPTION
**Pull request by tjazerzen**
This PR introduces two new OCaml modules, Node and Graph, along with an example script demonstrating their usage.

- **Node Module**: This module defines a Node structure, along with several associated functions. A Node is defined with an integer value, and functions are provided for creating a new Node, retrieving a Node's value and its ID, comparing two Nodes, and converting a Node to a string.

- **Graph Module**: The Graph module provides a structure for representing a graph and a set of functions for manipulating the graph. The Graph is represented as a map from Nodes to sets of Nodes (representing edges), and can be either directed or undirected. Functions are provided for creating an empty graph, adding and removing nodes, adding and removing edges, retrieving the list of nodes and edges in the graph, and converting the graph to a string.

- **graph_examples.ml**: This script provides a demonstration of the Node and Graph modules in action. It shows how to create Nodes, add them to a Graph, add edges between Nodes, and remove Nodes and edges from the Graph. It also demonstrates the use of the graph's to_string function to print a string representation of the graph.

Changes:

Add Node module with associated functions
Add Graph module with associated functions
Add example script graph_examples.ml

### CODE REVIEW

**Comment by tjazerzen** [comment]
Finished writing `Node` and `Graph` modules, which are written in `playground/graph/graph.ml` and used as examples in `playground/graph/graph_examples.ml`.

File: playground/graph/graph.exe
Comment (matijapretnar) [inline playground/graph/graph.exe:1]: Zakaj ta datoteka sploh obstaja? To bi morala biti knjižnjica, ne executable.

File: playground/graph/dune-project [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+(package
+ (name graph)
+ (synopsis "A short synopsis")
+ (description "A longer description")
+ (depends ocaml dune)
+ (tags
+  (topics "to describe" your project)))
+
+; See the complete stanza docs at https://dune.readthedocs.io/en/stable/dune-files.html#dune-project
```
Comment (matijapretnar) [inline playground/graph/dune-project]: Te vrstice niso potrebne, se jih lahko izbriše.

File: diploma/dokument-diplomska-naloga.pdf
Comment (matijapretnar) [inline diploma/dokument-diplomska-naloga.pdf:1]: Tudi PDF se lahko izbriše.

File: playground/graph/dune [hunk 2]
```diff
@@ -0,0 +1,5 @@
+(executable
+ (name graph_examples)
+ (promote
+  (until-clean))
+ (flags :standard -w -A))
```
Comment (matijapretnar) [inline playground/graph/dune]: Zakaj bi izklopili opozorila? Običajno so smiselna.

File: README.md [hunk 3]
```diff
@@ -1,3 +1,25 @@
...
+V okviru diplomske naloge bom raziskoval paralelizacijo grafovskih algoritmov v funkcijskih programskih jezikih.
+
+## Nastavitev okolja
+
+Za poganjanje projekta se zahteva package manager **opam** (inštalacija: https://opam.ocaml.org/doc/Install.html) ter knjižnice, zapisane v **requirements.txt**. Po naloženem opamu, poženite `opam install .` za nalaganje potrebnih modulov.
+
+
+## Struktura projekta
+
+V folderju /diploma so shranjeni pisni izdelki diplomske naloge (najpomembnejši izmed njih je verjetno [diplomska naloga sama](https://github.com/tjazerzen/parallelisation-of-graph-algorithms-in-functional-programming-languages/blob/master/diploma/dokument-diplomska-naloga.pdf)). Mapa /diploma uporablja rezultate iz mape /playground.
```
Comment (matijapretnar) [inline README.md]: `/diploma` in podobno za ostale poti.

File: playground/graph/graph_examples.exe
Comment (matijapretnar) [inline playground/graph/graph_examples.exe:1]: `.exe` sodi v `.gitignore`, ne v repozitorij.

File: README.md [hunk 4]
```diff
@@ -1,3 +1,25 @@
-# Parallelisation of Graph Algorithms in Functional Programming Languages
+# Paralelizacija grafovskih algoritmov v funkcijskih programskih jezikih
 
-In the scope of my bachelor's thesis I'll research parallelisation of graph algorithms in functional programming languages. Text is written in Slovene.
+V okviru diplomske naloge bom raziskoval paralelizacijo grafovskih algoritmov v funkcijskih programskih jezikih.
+
+## Nastavitev okolja
+
+Za poganjanje projekta se zahteva package manager **opam** (inštalacija: https://opam.ocaml.org/doc/Install.html) ter knjižnice, zapisane v **requirements.txt**. Po naloženem opamu, poženite `opam install .` za nalaganje potrebnih modulov.
+
```
Comment (matijapretnar) [inline README.md]: A se sploh rabi kakšen modul? Po mojem je `dune build` dovolj?

File: requirements.txt [hunk 5]
```diff
@@ -0,0 +1 @@
+"core" {>= "0.14.1"}
```
Comment (matijapretnar) [inline requirements.txt]: Kaj je to zaen paket?

File: playground/graph/graph_examples.ml [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+let my_graph = 
+  Graph.empty true
+  |> Graph.add_node node1 
+  |> Graph.add_node node2 
+  |> Graph.add_node node3
+  |> Graph.add_edge node1 node2
+  |> Graph.add_edge node2 node3
+
+
+let () = Printf.printf "Printing graph to string...\n"
```
Comment (matijapretnar) [inline playground/graph/graph_examples.ml]: To se da krajše napisati kot `print_endline "..."`

File: playground/graph/graph.ml [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+  let counter = ref 0
+
+  let create v = 
+    let id = !counter in
+    incr counter;
+    { id = id; value = v }
+
+  let value node = node.value
+
+  let id node = node.id
```
Comment (matijapretnar) [inline playground/graph/graph.ml]: Ali to funkcijo rabimo? Po mojem je bolje, da je id skrit.

File: playground/graph/graph.ml [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+end = struct
+  type elt = int
+  type t = { id : int; value : elt }
+  
+  let counter = ref 0
+
+  let create v = 
+    let id = !counter in
+    incr counter;
+    { id = id; value = v }
```
Comment (matijapretnar) [inline playground/graph/graph.ml]: Ker je ime vrednosti enako kot polje, se lahko na krajše piše kot `{ id; value = v }`. Če se argument `create` preimenuje iz `v` v `value`, velja podobno.

File: playground/graph/graph.ml [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+module NodeSet = Set.Make(Node)
+module NodeMap = Map.Make(Node)
+
+
+
+module Graph : sig
+  type elt = int  (* node values *)
+  type t    (* graphs *)
+  val empty : bool -> t
```
Comment (matijapretnar) [inline playground/graph/graph.ml]: Argument `directed` bi bilo dobro poimenovati (labelled arguments), da se ve, kaj pomeni.

File: playground/graph/graph.ml [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+  val add_node : Node.t -> t -> t
+  val remove_node : Node.t -> t -> t
+  val add_edge : Node.t -> Node.t -> t  -> t
+  val remove_edge : Node.t -> Node.t -> t -> t
+  val nodes : t -> Node.t List.t
+  val edges : t -> (Node.t * Node.t) list
```
Comment (matijapretnar) [inline playground/graph/graph.ml:40]: Koristna bi bila tudi funkcija `neighbours : t -> Node.t -> Node.t list`

File: playground/graph/graph.ml [hunk 11]
```diff
@@ -0,0 +1,118 @@
...
+
+module Graph : sig
+  type elt = int  (* node values *)
+  type t    (* graphs *)
+  val empty : bool -> t
+  val add_node : Node.t -> t -> t
+  val remove_node : Node.t -> t -> t
+  val add_edge : Node.t -> Node.t -> t  -> t
+  val remove_edge : Node.t -> Node.t -> t -> t
+  val nodes : t -> Node.t List.t
```
Comment (matijapretnar) [inline playground/graph/graph.ml]: Namesto `List.t` pišemo kar `list`.

File: playground/graph/graph.ml [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+  let remove_node node graph =
+    let edges =
+      graph.edges
+      |> NodeMap.remove node
+      |> NodeMap.map (NodeSet.remove node)
+    in
+    { graph with edges }
+
+  let add_edge node1 node2 graph =
+    let edges = graph.edges |> NodeMap.add node1 (NodeSet.add node2 (NodeMap.find node1 graph.edges)) in
```
Comment (matijapretnar) [inline playground/graph/graph.ml]: Ta del bi lahko dali v pomožno (in v signaturi skrito) funkcijo `add_undirected_edge`, ki doda samo enega, potem se pa v `add_edge` pokliče enkrat ali dvakrat. Podobno za `remove_edge`.

**Review by matijapretnar — CHANGES_REQUESTED** [review]
Odlično! Spodaj sem dodal nekaj manjših komentarjev, ampak načeloma tako naprej. Predlagam, da za konsistenten stil uporabite ocamlformat, lahko kar s privzetimi nastavitvami.

File: playground/graph/graph.exe
Comment (tjazerzen) [reply playground/graph/graph.exe:1]: Se strinjam, da. Odstranjujem...

Comment (tjazerzen) [reply playground/graph/graph.exe:1]: 71531ba7fad9ecee5d372c95943b0e05413461fd

File: playground/graph/dune-project (hunk 1 above)
Comment (tjazerzen) [reply playground/graph/dune-project]: 13db1c70940e5fe685463e3b42d1e20a24759ea2

File: diploma/dokument-diplomska-naloga.pdf
Comment (tjazerzen) [reply diploma/dokument-diplomska-naloga.pdf:1]: ef6b535044b19b83b61df4a04f0bcd8b96513848

File: playground/graph/dune (hunk 2 above)
Comment (tjazerzen) [reply playground/graph/dune]: 6a81f2282331ec751ab054b0ef97a5dddce4760d

File: playground/graph/dune-project (hunk 1 above)
Comment (tjazerzen) [reply playground/graph/dune-project]: 33b17144f280be33337d23e3493ea342fbdcaa06

File: requirements.txt (hunk 5 above)
Comment (tjazerzen) [reply requirements.txt]: Sorry, je ostalo od enih izmed prejšnih implementacij grafa, hvala da ste opazili. Odstranjujem...

File: README.md (hunk 4 above)
Comment (tjazerzen) [reply README.md]: Sem pozabil odstraniti `requirenments.txt` iz prejšnih implementacij grafa.
Resolved with b9c32cd0d7c41ebb0fba09e79cfbd1670f881bb3.

File: README.md (hunk 3 above)
Comment (tjazerzen) [reply README.md]: b9c32cd0d7c41ebb0fba09e79cfbd1670f881bb3.

File: playground/graph/graph_examples.exe
Comment (tjazerzen) [reply playground/graph/graph_examples.exe:1]: ab2b403ef178bbc659af8dbda4e8e00e64afccfd

File: requirements.txt (hunk 5 above)
Comment (tjazerzen) [reply requirements.txt]: b9c32cd0d7c41ebb0fba09e79cfbd1670f881bb3

File: playground/graph/graph_examples.ml (hunk 6 above)
Comment (tjazerzen) [reply playground/graph/graph_examples.ml]: 2730adeb273067d1eb6a882be53c2797132e0c21

File: playground/graph/graph.ml (hunk 7 above)
Comment (tjazerzen) [reply playground/graph/graph.ml]: 379652b20a5bc49e5fbc44c43ab795fafadd67ed

File: playground/graph/graph.ml (hunk 8 above)
Comment (tjazerzen) [reply playground/graph/graph.ml]: b73ffb0fdb47930c72d264ca3d36e2c237844a67

File: playground/graph/graph.ml (hunk 9 above)
Comment (tjazerzen) [reply playground/graph/graph.ml]: 8d23872a3b943b896e79c63239544bcccc2d5022

File: playground/graph/graph.ml (hunk 11 above)
Comment (tjazerzen) [reply playground/graph/graph.ml]: 31a93980fb7c464f41db51908403eb10d05e5b04

File: playground/graph/graph.ml (hunk 10 above)
Comment (tjazerzen) [reply playground/graph/graph.ml:40]: Dodajam. Metodo sem raje definiral kot `Node.t -> t -> Node.t list`, če bomo na nekemu grafu uporabljali operator `|>`.

fa4a1416377233a30f629932a4cede3aed78a95a

File: playground/graph/graph.ml (hunk 12 above)
Comment (tjazerzen) [reply playground/graph/graph.ml]: 7f3b11c4a2c4ec60ab3fa70b291a502b23cfc8a2

**Comment by tjazerzen** [comment]
Hvala za review :) Vaša navodila so bila precej "straightforward", vse potrebne spremembe pa zelo kratke, zato se mi zdi da ponoven code review ni potreben. Mergam.

**Comment by matijapretnar** [comment]
:+1:

**Pull request merged**